README.md
Disclaimer
==========

Fluff is still under heavy development. No guarantees are given ;). However, if you find it useful and encounter problems, let me know and I'll try to help. The gene annotation track that fluff_profile.py produces is still pretty ugly too.

fluff
=====

![fluff examples](https://raw.github.com/simonvh/fluff/master/examples/test.png) 

Fluff is a Python package containing several scripts with the aim to produce pretty, publication-quality figures for next-generation sequencing experiments. The goal is to provide default settings which produce figures that are ready-to-use (resolution, font size, colors etc.)

It currently contains three scripts:
* fluff_heatmap.py
* fluff_bandplot.py
* fluff_profile.py

Plotting is handled by the excellent [matplotlib](http://matplotlib.sourceforge.net/) library, several image formats are supported (SVG, Postscript, PDF, PNG).

Fluff makes heavy use of [pysam](http://code.google.com/p/pysam/), [pybedtools](http://packages.python.org/pybedtools/), [HTseq](http://www-huber.embl.de/users/anders/HTSeq/) and indexed [BAM files](http://samtools.sourceforge.net/) for speed and ease of use. Due to the indexing, the fluff scripts can be quick, even when working with large files. Currently, [BED](http://genome.ucsc.edu/FAQ/FAQformat.html#format1) files are also supported, but performance will suffer.

Prerequisites
-------------
* tabix - http://sourceforge.net/projects/samtools/files/
* pysam - http://code.google.com/p/pysam/
* pybedtools - http://packages.python.org/pybedtools/
* HTSeq - http://www-huber.embl.de/users/anders/HTSeq/
* matplotlib - http://matplotlib.sourceforge.net/
* numpy - http://numpy.scipy.org/
* scipy - http://www.scipy.org/
* colorbrewer - http://pypi.python.org/pypi/colorbrewer/
* Pycluster - https://pypi.python.org/pypi/Pycluster
* pp - http://www.parallelpython.com/ (optional)

I'd recommend installing matplotlib, numpy and scipy using your preferred package manager. Ubuntu, Debian, Fedora, Gentoo, etc. all have packages providing these libraries.

Colors
------
One important feature of fluff is the use of color. There are three ways to specify colors: by name, palette or hex code. You can specify as few or as many colors as you want, seperated by commas. If fluff runs out of colors it will start at the beginning.

1. **Color names.**
Fluff knowns nine color names: red, blue, green, purple, orange, yellow, pink and grey. These are the colors from the Set1 palette (see below).

2. **Palettes.**
The palettes with all the [ColorBrewer](http://colorbrewer2.org/) colors can be specified (thanks to the Python [colorbrewer](http://pypi.python.org/pypi/colorbrewer/) package supplied by Michael Hoffman). If you specify the palette name only, the maximum number of colors will be used. Alternatively specify the number of colors with a colon. For example: *Set1:5,Set2:4* is a valid argument.

3. **Hex code.**
Finally, colors can be specified using a hex code. For instance, *ff0000,00FF00,00000FF* will give you red, blue and green. Indeed, case-insensitive.

All different color specification can be mixed and matched for extra fun.

General options
---------------

There are three options shared amongst all scripts:
* `-r` Enable this option to use RPKM values (read per kb per million reads) instead of read counts.
* `-R` When bam files are produced by bwa, reads mapping to non-duplicate regions of the genome are marked. By default fluff removes all these reads. When this option is used these reads will be kept.
* `-D` By default fluff removes all duplicate reads (when marked in the bam file). Enable the `-D` option to keep duplicates.

The scripts
===========

fluff_heatmap.py
----------------
Produce a heatmap like [this example](add link). Features can be shown "as is", preserving the order in the input file, or can be clustered using hierarchical or k-means clustering. This scripts creates two output files. One is an image that contains the heatmap (png by default), the other one contains the features in the same order as the heatmap including cluster number. This file can be used as input for `fluff_bandplot.py`. 
It is worth mentioning that the heatmap can be save in a vector-based format, however, the file quickly becomes prohibitively large and unwieldy. Be warned ;).

```
Usage: fluff_heatmap.py -f <bedfile> -d <file1>[,<file2>,...] -o <out> [options]

Options:
  --version     Show program's version number and exit
  -h, --help    Show this help message and exit
  -f FILE       BED file containing features
  -d FILE(S)    Data files (reads in BAM or BED format)
  -o FILE       Output file (type determined by extension)

  Optional:
    -C METHOD   kmeans, hierarchical or none
    -k INT      Number of clusters
    -m          Merge mirrored clusters (only with kmeans)
    -c NAME(S)  Color(s) (name, colorbrewer profile or hex code)
    -B NAME(S)  Background color(s) (name, colorbrewer profile or hex code)
    -e INT      Extend (in bp. Default: 5000)
    -b INT      Bin size (default 100)
    -s SCALE    Scale (absolute or percentage)
    -r          Use RPKM instead of read counts
    -D          Keep duplicate reads (removed by default)
    -R          Keep repeats (removed by default, bwa only)
    -O          Save read count matrix
    -P          Define number of CPUs
    -M METHOD   Euclidean or pearson
    -g          Identify dynamics by extending features 1kb up/down stream(just for clustering), cluster as 1 bin, diplay as original number of bins and with the default extend values
```

### Clustering ###
By default, fluff_heatmap.py will preserve the order of the features in the input BED file. This is equivalent to specifying `-C none`. Alternatively, one of two basic clustering methods can be specified using the `-C` parameter: `hierarchical` and `kmeans`. If `kmeans` is selected the number of clusters (`-k`) is mandatory. The profiles will be clustered using the Euclidian distance metric, based on normalized data (although the heatmap still shows the raw data). If the regions in the input BED file are not strand-specific (peaks, for instance), different clusters might describe the same strand-specific profile in two different orientations. In this case you can select the `-m` parameter to merge clusters which are similar, but mirrored around the center coordinate. Similarity is based on the chi-squared p-value of the mean profile per cluster. **All** tracks have to pass this similarity test for two clusters to be merged! Merging follows a iterative, greedy procedure where the two most similar clusters are merged after which the process is repeated.

### Region, bin size and scaling ###
The *center* of all features will be selected from the input BED file. A total region of 10kb will be centered at this coordinate (5kb up- and downstream). To change this, specify the `-e` parameter that controls the extension up- and downstream. For example,  `-e 10000` will select regions of 20kb.

The bin size is 100 bp, this can be controlled with the `-b` parameter.

The color scale of the heatmap profile can be adjusted with the `-s` parameter. It is difficult to set a generally applicable threshold, as this value might depend on what you'd like to highlight or analyze in your data. The `-s` argument can either be absolute (`-s 15` means a bin that contains 15 or more reads will have the darkest color) or a percentage (`-s 90%` sets the threshold at the 90th percentile of the values in all bins over all tracks and features).

### Background color ###
The default background color is white. This can be changed with the `-B` parameter, which works the same as the color specification.

fluff_bandplot.py
-----------------
While heatmaps can be very informative, sometimes you want to show the average profile. However, by just plotting the mean, you lose quite some information. This is my attempt to combine a boxplot with an average profile. The mean enrichment is shown using a black line. The 50th and 90th percentile are also visualized using a dark and light color respectively.

```
Usage: fluff_bandplot.py -c <bedfile> -d <file1>[,<file2>,...] -o <out> [options]

Options:
  --version     Show program's version number and exit
  -h, --help    Show this help message and exit
  -i FILE       BED file with cluster in 4th column
  -d FILE(S)    Data files (reads in BAM or BED format)
  -o FILE       Output file (type determined by extension)

  Optional:
    -c NAME(S)  Color(s) (name, colorbrewer profile or hex code)
    -s GROUPS   Scale groups
    -p INT,INT  Range of percentiles (default 50,90)
    -r          Use RPKM instead of read counts
    -D          Keep duplicate reads (removed by default)
    -R          Keep repeats (removed by default, bwa only)
```

The input file should contain a cluster identifier in the fourth column. Any string or number can be used; all features with the same identifier will be grouped. The size of the regions is determined by the coordinates in the BED file, ie if you want to plot a region of 10kb all the entries in the BED file should span a region of 10kb. The `*_clusters.bed` file from `fluff_heatmap.py` can directly be used in this script.

### Scales and percentiles ###
To compare different datasets (for instance same antibody in a timeline or different conditions) the maximum of the Y-axis can be linked between tracks. As an example, `-s 1:3,4,5:7` will ensure that the first three rows share the same scale, the fourth track will be scaled to the maximum of only this single track and the last three rows will once again share a similar scale.

The `-p` parameter controls the percentile range. By default, the profile that contains 50% of the data will be shaded in a dark color and the profile that contains 90% of the data will be shaded in a lighter color. 

fluff_profile.py
----------------
Produces output like a Genome Browser screenshot. Currently only 1) profiles based on reads in BAM or BED format and 2) gene annotation (in [BED12](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)) can be visualized.
```
Usage: fluff_profile.py -i <loc1>[,<loc2>,...] -d <file1>[,<file2>,...] -o <out> [options]

Options:
  --version           Show program's version number and exit
  -h, --help          Show this help message and exit
  -i INTERVAL(S)      One or more genomic intervals (chrom:start-end)
  -d FILE(S)          Data files (reads in BAM or BED format)
  -o FILE             Output file name (type determined by extension)

  Optional:
    -a FILE           Annotation in BED12 format
    -c NAME(S)        Color(s) (name, colorbrewer profile or hex code)
    -t GROUPS         Track groups
    -s GROUPS         Scale groups
    -S SCALE          Scale: 'auto' (default), 'off' or int for each track
    -b BACKGROUND     Background color: white | color | stripes
    -f FRAGMENTSIZE   Fragment length (default: 200)
```

###Scale and track groups###
Just as with `fluff_bandplot.py`, identical Y-axis scale can be used to compare tracks. In addition, different datasets can be overlayed on the same track wit the `-t` option. Use the `-S` option to manually set a the Y-axis upper limit, or to turn the label off (with the `off` argument).

###Colors###
The background color can be one of three different choices. The default is `white`. Specifying `color` will set the background of each track to a lighter version of the track color. The last option is `stripes` which will show an alternating pattern of white and light grey bands.

###Fragment length###
By default, all reads will be extended to 200bp before creating the profiles. Set an alternative fragment size with the `-f` option. Paired-end profiles are not yet supported, paired reads will be independently processed.







ChangeLog
Fluff

Version : 1.42
Date : 26 02 2014

script/
	fluff_heatmap.py :
		Pearson as distance function.
		Multiprocessing.
		Try to identify dynamics by extending 1 kb up/down stream, just for clustering
		and display the data with the default settings
	Warning when using too many data (Issue: #17).
		If there is a gene name in BED file will be added to
		'_clusters.bed' file (Issue: #16).
		Get reads matrix count as an output ('_readcount.txt') (Issue:
		#18).
		Fixed bugs.
		Fixed clusters numbers
	fluff_bandplot.py :
		"_clusters.bed" files with/without gene name can be used to develop bandplots
		-P option added to chance the the percentile
	fluff_profile.py : 

fluff/
	fluffio.py :
		If there is a gene name in BED file will be added to 
		'_clusters.bed' file (Issue: #16).
		Fixed bugs.
	color.py : 
	config.py :
		Global version added.
	plot.py : 
	util.py : 
	


scripts/fluff_heatmap.py
#!/usr/bin/env python
# Copyright (c) 2012-2013 Simon van Heeringen <s.vanheeringen@ncmls.ru.nl>
#
# This script is free software. You can redistribute it and/or modify it under 
# the terms of the MIT License

### Standard imports ###
from optparse import OptionParser,OptionGroup
from tempfile import NamedTemporaryFile
import sys
import os
import os.path
import multiprocessing

### External imports ###
from numpy import array,hstack,arange,median,mean,zeros
from scipy.stats.mstats import rankdata
import Pycluster
from math import sqrt,log

### My imports ###
from fluff.util import *
from fluff.fluffio import *
from fluff.color import DEFAULT_COLORS, parse_colors
from fluff.plot import heatmap_plot
from fluff.config import *

VERSION = str(FL_VERSION)

DEFAULT_BINSIZE = 100
DEFAULT_METRIC = "e"        # Euclidian, PyCluster
FONTSIZE = 8
DEFAULT_SCALE = "90%" 
DEFAULT_EXTEND = 5000
DEFAULT_PERCENTILE = 99
DEFAULT_CLUSTERING = "none"
DEFAULT_BG = "white"
DEFAULT_DYN_EXTEND = 5000

usage = "Usage: %prog -f <bedfile> -d <file1>[,<file2>,...] -o <out> [options]"
version = "%prog " + str(VERSION)
parser = OptionParser(version=version, usage=usage)
group1 = OptionGroup(parser, 'Optional')

parser.add_option("-f", 
                  dest="featurefile", 
                  help="BED file containing features", 
                  metavar="FILE")
parser.add_option("-d", 
                  dest="datafiles", 
                  help="data files (reads in BAM or BED format)", 
                  metavar="FILE(S)")
parser.add_option("-o", 
                  dest="outfile", 
                  help="output file (type determined by extension)", 
                  metavar="FILE")

# Optional arguments
group1.add_option("-p", 
                  dest="pick", 
                  help="pick specific data files to use for clustering", 
                  default=None,
                  type="string")
group1.add_option("-C", 
                  dest="clustering", 
                  help="kmeans, hierarchical or none", 
                  default=DEFAULT_CLUSTERING, 
                  metavar="METHOD")
group1.add_option("-k", 
                  dest="numclusters", 
                  help="number of clusters", 
                  metavar="INT", 
                  type="int", 
                  default=1)
group1.add_option("-m", 
                  dest="merge_mirrored", 
                  help="merge mirrored clusters (only with kmeans and without -g option)", 
                  metavar="", 
                  default=False, 
                  action="store_true")
group1.add_option("-c", 
                  dest="colors", 
                  help="color(s) (name, colorbrewer profile or hex code)", 
                  metavar="NAME(S)", 
                  default=DEFAULT_COLORS)
group1.add_option("-B", 
                  dest="bgcolors", 
                  help="background color(s) (name, colorbrewer profile or hex code)", 
                  metavar="NAME(S)", 
                  default=DEFAULT_BG)
group1.add_option("-e", 
                  dest="extend", 
                  help="extend (in bp. Default: {0})".format(DEFAULT_EXTEND), 
                  metavar="INT", 
                  type="int", 
                  default=DEFAULT_EXTEND)
group1.add_option("-b", 
                  dest="binsize", 
                  help="bin size (default {0})".format(DEFAULT_BINSIZE), 
                  metavar="INT", 
                  type="int", 
                  default=DEFAULT_BINSIZE)
group1.add_option("-s", 
                  dest="scale",
                  help="scale (absolute or percentage)", 
                  metavar="", 
                  type="string", 
                  default=DEFAULT_SCALE)
group1.add_option("-F", 
                  dest="fragmentsize", 
                  help="Fragment length (default: read length)",
                  type="int",
                  default=None)
group1.add_option("-r", 
                  dest="rpkm", 
                  help="use RPKM instead of read counts", 
                  metavar="", 
                  action="store_true", 
                  default=False)
group1.add_option("-D", 
                  dest="rmdup", 
                  help="keep duplicate reads (removed by default)", 
                  metavar="", 
                  default=True, 
                  action="store_false")
group1.add_option("-R", 
                  dest="rmrepeats", 
                  help="keep repeats (removed by default, bwa only) ", 
                  metavar="", 
                  action="store_false", 
                  default=True)
group1.add_option("-P", 
                  dest="cpus", 
                  help="number of CPUs (default: 4)", 
                  metavar="INT", 
                  type="int", 
                  default=4)
group1.add_option("-M", 
                  dest="distancefunction", 
                  help="Euclidean or Pearson (default: Euclidean)", 
                  default=DEFAULT_METRIC,
                  metavar="METHOD")
group1.add_option("-g", 
                  dest="graphdynamics", 
                  help="Identify dynamics by extending features 1kb up/down stream(just for clustering), cluster as 1 bin, diplay as original number of bins and with the default extend values",  
                  metavar="", 
                  action="store_true", 
                  default=False)
parser.add_option_group(group1)
(options, args) = parser.parse_args()

#Check if files exist
for opt in [options.featurefile, options.datafiles, options.outfile]:
    if not opt:
        parser.print_help()
        sys.exit()
if not os.path.isfile(options.featurefile):
  print "ERROR: The BED files which contains the features file does not exist!"
  sys.exit(1)
for x in options.datafiles.split(","):
  if not os.path.isfile(x):
    print "ERROR: Data file '{0}' does not exist".format(x)
    sys.exit(1)
for x in options.datafiles.split(","):
  if '.bam' in x and not os.path.isfile("{0}.bai".format(x)):
    print "Data file '{0}' does not have an index file".format(x)
    print "Creating an index file for {0}".format(x)
    pysam.index(x)
    print "Done!"
#Options Parser
featurefile = options.featurefile
datafiles = [x.strip() for x in options.datafiles.split(",")]
tracks = [os.path.basename(x) for x in datafiles]
titles = [os.path.splitext(x)[0] for x in tracks]
colors = parse_colors(options.colors)
bgcolors = parse_colors(options.bgcolors)
outfile = options.outfile
extend_up = options.extend
extend_down = options.extend
fragmentsize = options.fragmentsize
cluster_type = options.clustering[0].lower()
merge_mirrored = options.merge_mirrored
bins = (extend_up + extend_down) / options.binsize
rmdup = options.rmdup
rpkm = options.rpkm
rmrepeats = options.rmrepeats
ncpus = options.cpus
distancefunction = options.distancefunction[0].lower()
dynam = options.graphdynamics

#Check for given number of processors
if (ncpus>multiprocessing.cpu_count()):
  print "ERROR: You can use only up to {0} processors!".format(multiprocessing.cpu_count())
  sys.exit(1)
#Check for mutually exclusive parameters
if merge_mirrored and dynam:
  print "ERROR: -m and -g option CANNOT be used together"
  sys.exit(1)
#Warning about too much files
if (len(tracks)>4):
  print "Warning: Running fluff with too many files might make you system use enormous amount of memory!"
#Method of clustering
if (options.pick != None):
  pick = [i - 1 for i in split_ranges(options.pick)]
else:
  pick = range(len(datafiles))
if not cluster_type in ["k", "h", "n"]:
    sys.stderr.write("Unknown clustering type!\n")
    sys.exit(1)
#Number of clusters
if cluster_type == "k" and not options.numclusters >= 2:
    sys.stderr.write("Please provide number of clusters!\n")
    sys.exit(1)
#Distance function
if not distancefunction in ["e", "p"]: 
  sys.stderr.write("Unknown distance function!\n")
  sys.exit(1)
else:
  if distancefunction == "e":
    METRIC = DEFAULT_METRIC
    print "Euclidean distance method"
  else:
    METRIC = "c"
    print "Pearson distance method"
## Get scale for each track
tscale = [1.0 for track in datafiles]

#Function to load heatmap data
def load_data(featurefile, amount_bins, extend_dyn_up, extend_dyn_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard=[]):
  # Calculate the profile data
  data = {}
  regions = []
  if guard or not dynam:
    print "Loading data"
  try:
    # Load data in parallel
    import pp
    job_server = pp.Server(ncpus)
    jobs = []
    for datafile in datafiles:
        jobs.append(job_server.submit(load_heatmap_data, (featurefile, datafile, amount_bins, extend_dyn_up, extend_dyn_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard),  (), ("tempfile","sys","os","fluff.fluffio","numpy")))
    for job in jobs:
        track,regions,profile,guard = job()
        data[track] = profile
  except:
    sys.stderr.write("Parallel Python (pp) not installed, can't load data in parallel\n")
    for datafile in datafiles:
        track,regions,profile,guard = load_heatmap_data(featurefile, datafile, amount_bins, extend_dyn_up, extend_dyn_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard)
        data[track] = profile
  return data, regions, guard

#-g : Option to try and get dynamics
#Extend features 1kb up/down stream
#Cluster them in one bin
guard=[]
if dynam:
  amount_bins = 1
  extend_dyn_up = 1000
  extend_dyn_down = 1000
  data, regions, guard = load_data(featurefile, bins, extend_up, extend_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard)
else:
  amount_bins = bins
  extend_dyn_up = extend_up
  extend_dyn_down = extend_down

#Load data for clustering
data, regions, guard = load_data(featurefile, amount_bins, extend_dyn_up, extend_dyn_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard)
# Normalize
norm_data = normalize_data(data, DEFAULT_PERCENTILE)
clus = hstack([norm_data[t] for i,t in enumerate(tracks) if (not pick or i in pick)])

#Clustering
if cluster_type == "k":
    print "K-means clustering"
    ## K-means clustering
    # PyCluster
    labels, error, nfound = Pycluster.kcluster(clus, options.numclusters, dist=METRIC)
    if not dynam and merge_mirrored:
        (i,j) = mirror_clusters(data, labels)
        while j:
            for track in data.keys():
                data[track][labels == j] = [row[::-1] for row in data[track][labels == j]]
            for k in range(len(regions)):
                if labels[k] == j:
                    (chrom,start,end,gene,strand) = regions[k]
                    if strand == "+":
                        strand = "-"
                    else:
                        strand = "+"
                    regions[k] = (chrom, start, end, gene, strand)
            n = len(set(labels))
            labels[labels == j] = i
            for k in range(j + 1, n):
                labels[labels == k] = k - 1
            (i,j) = mirror_clusters(data, labels)
            
    ind = labels.argsort()
    # Other cluster implementation
    #    centres, labels, dist = kmeanssample(clus, options.numclusters, len(clus) / 10,  metric=cl, maxiter=200, verbose=1, delta=0.00001)
elif cluster_type == "h":
    print "Hierarchical clustering"
    tree = Pycluster.treecluster(clus, method="m", dist=METRIC)
    labels = tree.cut(options.numclusters)
    ind = sort_tree(tree, arange(len(regions)))
else:
    ind = arange(len(regions))
    labels = zeros(len(regions))
f = open("{0}_clusters.bed".format(outfile), "w")
for (chrom,start,end,gene,strand), cluster in zip(array(regions, dtype="object")[ind], array(labels)[ind]):
  if not gene:
    f.write("{0}\t{1}\t{2}\t.\t{3}\t{4}\n".format(chrom, start, end, cluster+1, strand))
  else: 
    f.write("{0}\t{1}\t{2}\t{3}\t{4}\t{5}\n".format(chrom, start, end, gene, cluster+1, strand))
f.close()

if not cluster_type == "k":
    labels = None

#Save read count matrix
input_file = open('{0}_readcount.txt'.format(outfile), 'w')
for i, track in enumerate(tracks):
  for k, v in data.items():
    if track == k:
      order = []
      for row in v:
        for x in row:
          order.append(x)
      for j in ind:
        input_file.write('{0}\t'.format(str(order[j])))
      input_file.write('\n')

#Load data for visualization if -g option was used
if dynam:
  data, regions, guard = load_data(featurefile, bins, extend_up, extend_down, rmdup, rpkm, rmrepeats, fragmentsize, dynam, guard)

scale = get_absolute_scale(options.scale, [data[track] for track in tracks])
heatmap_plot(data, ind, outfile, tracks, titles, colors, bgcolors, scale, tscale, labels)
